harvdev_utils/examine_archives/search_archives.py

```diff
@@ -0,0 +1,133 @@
+"""Search archives for specific fields WITH values."""
+import re
+import argparse
+import subprocess
+help = """   Start with most recent and search backwards.
+
+   Required field name: (-f --field)
+     OR
+   Required value: (-v --value)
+
+   Option to find a certain number of these.
+     Default is 1. (-l --limit)
+
+   Option to search just cambridge OR harvard.
+     Default both. (-s --site)
+
+   # Later maybe Option to lookup specific value too.
+
+   Usage examples
+   python search_archives.py -l 5 -f GA2c
+
+   python search_archives.py -v Oseg3
+
+"""
+parser = argparse.ArgumentParser(formatter_class=argparse.RawDescriptionHelpFormatter)
+parser.add_argument('-l', '--limit', help='Limit to x finds', type=int, default=1, required=False)
+parser.add_argument('-s', '--site', help="Restrict to harvard or cambridge", default=None, required=False)
+parser.add_argument('-f', '--field', help="Proforma filed name", default=None, required=False)
+parser.add_argument('-v', '--value', help="value, symbol name etc", default=None, required=False)
+parser.add_argument('-d', '--debug', help="For debugging NOT recomended", default=False, required=False)
+
+args = parser.parse_args()
+
+cam_start = '/data/camdata/proforma/'
+harv_start = '/data/harvcur/archive/'
+debug = False
+if args.debug:
+    debug = True
+field_search = True
+if not (args.field or args.value):
+    print("field or value is required")
+    print(help)
+    exit(-1)
+if args.field and args.value:
+    print("field OR value is required. Not both")
+    exit(-1)
+
+process_cam = False
+process_harv = False
+if args.site:
+    if args.site == 'cambridge':
+        process_cam = True
+    elif args.site == 'harvard':
+        process_harv = True
+    else:
+        print("Site must be unspeicified OR cambridge or harvard NOT {}".format(args.site))
+else:
+    process_cam = True
+    process_harv = True
+
+
+def get_ordered_directory_list(base_dir):
+    """Get list of directorys to examine.
+
+    This is the same for cambridge as harvard so does not matter which.
+    """
+    dir_list = []
+    cmd = "ls -lt {}".format(base_dir)
+    if debug:
+        print(cmd)
+    output = subprocess.getoutput(cmd)
+    pattern = r"\d{4}_\d{2}"
+
+    for line in output.split('\n'):
+        if debug:
+            print(line)
+        filename = line.split()[-1].strip()
+        if re.match(pattern, filename):
+            dir_list.append(filename)
+    return dir_list
+
+
+def search_proforma(grep_dir_string, field=None, value=None):
+    """Lookup field or value in the files represented by grep_dir_string."""
+    if field:
+        cmd = "grep '!.{}' {}".format(field, grep_dir_string)
```

Comment:
Should there be a period between the `!` and the `{}` bits? There's usually a space between the '!' and the field abbreviation in proforma fields. When I try to grep with the `.` in there, I get a message like `.G1a: Event not found`.

Reply:
'.' means 1 char of anything. SO could be a space or c.
grep '!.MP18' /data/camdata/proforma//2050_04/*/*/*
/data/camdata/proforma//2050_04/epicycle7/PHASE1/pl179402.bibl:! MP18. Action - make this/these multipub ID(s) secondary IDs of MP1  *N :
/data/camdata/proforma//2050_04/epicycle7/PHASE1/pl179402.bibl:! MP18. Action - make this/these multipub ID(s) secondary IDs of MP1  *N :

But that could be system dependent?

Ah does !c still has a space so those would be missed. I will merge anyway and fix this afterwards.

Reply:
Hmm. I've tested on `abe` and `flysql20` - I need to add a backslash before the exclamation point to avoid the "Event not found" message: i.e., `grep '\!.G1a' test`.
Also, if I use `*` instead of `.`, it catches the "!c" instances: i.e., `grep '\!*G1a' /data/camdata/proforma/2021_03/**`

Reply:
But on my Ubuntu at home, I don't need that extra backslash. Any, for what it's worth ...

Reply:
The problem with the  '*' syntax is it will catch things like ...
! GA10c. Associated insertion - G1a is outwith insert      *G :
Etc so I resisted the less strict try.

Reply:
I will try the '\' at the start though.